### PR TITLE
Add spacy word vectors to Doc

### LINKED ISSUE #12: Create Document Embeddings

### PR DESCRIPTION
**Pull request by KleinRana**
This exposes spacy word vectors as a property of `Doc`. This closes #12.

Note that the Spacy models Textpipe typically uses (e.g. `en_core_web_sm` and `nl_core_news_sm`) do not contain vectors. Hence the generated vectors are [context-sensitive tensors](https://github.com/explosion/spaCy/issues/2523#issuecomment-403453356).

We are thinking of adding other word embeddings in a separate pull request, maybe based on [fastText](https://github.com/facebookresearch/fastText/) or [BERT](https://github.com/google-research/bert). We left an `Operator` implementation to add this to a `Pipeline` for when we have more general word vectors.

### CODE REVIEW

File: textpipe/doc.py [hunk 1]
```diff
@@ -14,13 +14,13 @@
...
 
+from textpipe.data.emoji import emoji2unicode_name, emoji2sentiment
+
 
 class TextpipeMissingModelException(Exception):
     """Raised when the requested model is missing"""
     pass
 
-from textpipe.data.emoji import emoji2unicode_name, emoji2sentiment
-
```
Comment (KleinRana) [inline textpipe/doc.py]: Unrelated formatting change.

File: textpipe/doc.py [hunk 2]
```diff
@@ -375,3 +375,35 @@ def keyterms(self):
...
+        >>> doc.vectors['Test'][3].shape[-1]
+        384
+        >>> doc.vectors['Test'][1] == doc.vectors['sentence'][1]
+        False
+        """
+        return {token.text: (token.has_vector,
+                             token.vector_norm,
+                             token.is_oov,
+                             token.vector)
+                for token in self._spacy_doc}
```
Comment (fredericklu) [inline textpipe/doc.py]: If we are still using the `self._spacy_doc` object here, it will still use the spaCy's small models currently (e.g. `en_core_web_sm` and `nl_core_news_sm` as indicated).

What we did instead in our SJTextpipe is that we applied a new spaCy nlp model to the `nlp(doc.clean)` for the word embeddings operation. That will give us the real vectors. External word embedding models are based on [fastText ](https://github.com/facebookresearch/fastText/blob/master/docs/crawl-vectors.md). Check the instruction here: https://spacy.io/usage/vectors-similarity#converting

File: textpipe/doc.py [hunk 3]
```diff
@@ -375,3 +375,35 @@ def keyterms(self):
...
+
+    @property
+    def vectors(self):
+        """
+        Returns word embeddings for the words in the document.
+        The used spacy models don't have "true" word vectors
+        but only context-sensitive tensors that are within the document.
+
+        Returns:
+        A dictionary mapping words from the document to a 4-tuple with the
```
Comment (anneschuth) [inline textpipe/doc.py]: Why do we choose to use a 4 tuple? I guess a dict could be more descriptive.

Comment (lmdehaas) [reply textpipe/doc.py]: I'd prefer a named tuple over a dict; we could just copy the SpaCy naming (`has_vector`, `vector_norm`, etc) for transparency. This would result in this if you want to retrieve a vector: `doc.vectors['sentence'].vector`

File: textpipe/doc.py [hunk 4]
```diff
@@ -375,3 +375,35 @@ def keyterms(self):
...
+        corresponding values of the following variables:
+
+        has vector: Does the token have a vector representation?
+        vector norm: The L2 norm of the token's vector (the square root of the
+                    sum of the values squared)
+        OOV: Out-of-vocabulary (This variable always gets the vlue TRUE since
```
Comment (lmdehaas) [inline textpipe/doc.py]:
```suggestion
        OOV: Out-of-vocabulary (This variable always gets the value True since
```

File: textpipe/doc.py [hunk 5]
```diff
@@ -375,3 +375,35 @@ def keyterms(self):
...
+                                there are no vectors included in the model)
+        vector: The vector represenation of the word
```
Comment (lmdehaas) [inline textpipe/doc.py]:
```suggestion
        vector: The vector representation of the word
```

File: textpipe/doc.py (hunk 3 above)
Comment (lmdehaas) [reply textpipe/doc.py]: > What would that look like in serialized form?

Only option is to serialize it as a dict is guess (through `as_dict()`)

**Comment by lmdehaas** [comment]
One more thing: I would like to see an operation corresponding to this property.

File: textpipe/doc.py (hunk 3 above)
Comment (dodijk) [reply textpipe/doc.py]: Maybe a dict makes a bit more sense then? We have a few more places were we return a tuple and a (list of) dict(s) makes more sense (`ents`, `emojis`, etc.). Should we add that as an issue?

File: textpipe/doc.py (hunk 2 above)
Comment (dodijk) [reply textpipe/doc.py]: Sounds pretty solid. Could you share your code for this with at least @KleinRana? Then we can come up with a plan to consolidate this.

File: textpipe/doc.py (hunk 3 above)
Comment (KleinRana) [reply textpipe/doc.py]: A `namedtuple` serializes as list by default: 
```python
>>> from collections import namedtuple
>>> Point = namedtuple('Point', ['x', 'y'])
>>> import json
>>> json.dumps(Point(1, 2))
'[1, 2]'
```
We will represent this as a `dict`, not as a `namedtuple`.

File: textpipe/doc.py [hunk 6]
```diff
@@ -413,3 +414,84 @@ def similarity(self, other_doc, metric='jaccard', hash_method='minhash'):
...
+        >>> doc = Doc('sentence with an out of vector word lsseofn')
+        >>> doc.aggregate_word_vectors().shape
+        (384,)
+        >>> numpy.array_equiv(doc.aggregate_word_vectors(exclude_oov=False), doc.aggregate_word_vectors(exclude_oov=True))
+        False
+        """
+        if exclude_oov:
+            tokens = [token for token in self._spacy_doc if not token.is_oov]
+        else:
+            tokens = [token for token in self._spacy_doc]
```
Comment (anneschuth) [inline textpipe/doc.py]: the above 4 lines could be 
```suggestion
            tokens = [token for token in self._spacy_doc if not exclude_oov or not token.is_oov]
```

File: textpipe/doc.py [hunk 7]
```diff
@@ -413,3 +414,84 @@ def similarity(self, other_doc, metric='jaccard', hash_method='minhash'):
...
+
+        if normalize:
+            vectors = [token.vector/token.vector_norm for token in tokens]
+        else:
+            vectors = [token.vector for token in tokens]
```
Comment (anneschuth) [inline textpipe/doc.py]: same here:
```suggestion
            vectors = [token.vector/token.vector_norm if normalize else token.vector for token in tokens]
```

File: textpipe/doc.py (hunk 6 above)
Comment (KleinRana) [reply textpipe/doc.py]: I agree that's indeed nicer

File: textpipe/doc.py [hunk 8]
```diff
@@ -431,15 +434,64 @@ def vectors(self):
...
+
+        if aggregation == 'mean':
+            return numpy.mean(vectors, axis=0)
+        elif aggregation == 'sum':
+            return numpy.sum(vectors, axis=0)
```
Comment (lmdehaas) [inline textpipe/doc.py]: Do we know how this serializes?

File: textpipe/doc.py [hunk 9]
```diff
@@ -431,15 +434,64 @@ def vectors(self):
...
         False
         """
-        return {token.text: (token.has_vector,
-                             token.vector_norm,
-                             token.is_oov,
-                             token.vector)
+        return {token.text: {'has_vector': token.has_vector,
+                             'vector_norm': token.vector_norm,
+                             'is_oov': token.is_oov,
+                             'vector': token.vector}
```
Comment (lmdehaas) [inline textpipe/doc.py]: What's the type of the vector, and does this type serialize and deserialize to and from JSON?

File: textpipe/doc.py [hunk 10]
```diff
@@ -413,3 +414,78 @@ def similarity(self, other_doc, metric='jaccard', hash_method='minhash'):
...
+        >>> numpy.array_equiv(Doc('a b').doc_vector, Doc('a b').doc_vector)
+        True
+        >>> numpy.array_equiv(Doc('a b').doc_vector, Doc('a a b').doc_vector)
+        False
+        """
+        return self.aggregate_word_vectors()
+
+    def aggregate_word_vectors(self, aggregation='mean', normalize=False, exclude_oov=False):
+        """
+        Returns document embeddings based on the words in the document.
```
Comment (graus) [inline textpipe/doc.py]: A more general comment: in our use-case, we want to compute multiple aggregations of a doc, to be precise:

```
        vector_list = [token.vector for token in doc_vectors]
        doc_vectors_mean = doc_vectors.vector.tolist()  # the same as np.mean(vector_list, axis=0)
        doc_vectors_sum = np.sum(vector_list, axis=0).tolist()
        doc_vectors_var = np.var(vector_list, axis=0).tolist()
        doc_vectors_l2norm = doc_vectors.vector_norm

        return {'word2vec_average': doc_vectors_mean,
                'word2vec_sum': doc_vectors_sum,
                'word2vec_variation': doc_vectors_var,
                'word2vec_l2norm': doc_vectors_l2norm}
```

In the current implementation, that would mean redoing the aggregation for each --- something to consider?

File: textpipe/doc.py (hunk 9 above)
Comment (KleinRana) [reply textpipe/doc.py]: The vectors are ndarrays which are not directly serializable to JSON with a JSON dump directly. I'll change the code such that they will be returned as lists (by using `.tolist()`). In this way they can be serialized to JSON and deserialised by changing them back to numpy arrays (e.g. for the doc vector: `numpy.array(json.dumps(doc.vector.tolist()))`.

File: textpipe/doc.py (hunk 10 above)
Comment (dodijk) [reply textpipe/doc.py]: I don't think generating the vector_list is the slowest thing here. It's the mean, sum and var operations. We could benchmarkt this, but I would opt too leave as is. We could at the var operation, if you are actually using that.

File: textpipe/doc.py (hunk 2 above)
Comment (KleinRana) [reply textpipe/doc.py]: I will change self._spacy_doc to self._load_spacy_doc(lang, model_name) such that external models can be used
